tensorboard/plugins/hparams/list_session_groups.py

```diff
@@ -0,0 +1,393 @@
+                                        value=metric_evals[-1][2]))
+    return result
+
+  def _build_group_from_sessions(self, sessions, session_infos_by_name):
+    assert len(sessions) > 0
+    # TODO(erez): Do proper metric aggregation. For now we just take
+    # the metric values from the first session.
+    result = api_pb2.SessionGroup(
+        name=session_infos_by_name[sessions[0].name].start_info.group_name,
+        metric_values=sessions[0].metric_values,
+        # Sort sessions by name so the order is deterministic.
+        sessions=sorted(sessions, key=lambda session : session.name),
+        monitor_url=(
+            session_infos_by_name[sessions[0].name].start_info.monitor_url
+        )
+    )
+    # Copy hparams from the first session (all sessions should have the same
+    # hyperparameter values) into result.
+    # There doesn't seem to be a way to initialize a protobuffer map in the
+    # constructor.
+    hparams = session_infos_by_name[sessions[0].name].start_info.hparams
+    for (key, value) in six.iteritems(hparams):
+      result.hparams[key].CopyFrom(value)
+    return result
+
+  def _filter(self, session_groups):
+    return filter(
+        lambda sg : self._passes_all_filters(sg),
+        session_groups)
+
+  def _passes_all_filters(self, session_group):
+    for f in self._filters:
+      if not f.passes(session_group):
+        return False
+    return True
+
+  def _sort(self, session_groups):
+    """Sorts 'session_groups' in place according to _request.col_params"""
+    # Sort by session_group name so we have a deterministic order.
+    session_groups.sort(key=lambda session_group : session_group.name)
+    # Sort by lexicographical order of the _request.col_params whose order
+    # is not ORDER_UNSPECIFIED. The first such column is the primary sorting
+    # key, the second is the secondary sorting key, etc. To achieve that we
+    # need to iterate on these columns in reverse order (thus the primary key
+    # is the key used in the last sort).
+    for col_param, extractor in reversed(zip(self._request.col_params,
+                                             self._extractors)):
+      if col_param.order == api_pb2.ORDER_UNSPECIFIED:
+        continue
+      key_func = lambda sg : extractor.extract(sg)
+      if col_param.order == api_pb2.ORDER_ASC:
+        session_groups.sort(key=key_func)
+      elif col_param.order == api_pb2.ORDER_DESC:
+        session_groups.sort(key=key_func, reverse=True)
+      else:
+        raise error.HParamsError('Unknown col_param.order given: %s' %
+                                 col_param)
+
+  def _create_response(self, session_groups):
+    return api_pb2.ListSessionGroupsResponse(
+        session_groups=session_groups[
+            self._request.start_index:
+            self._request.start_index+self._request.slice_size],
+        total_size=len(session_groups))
+
+
+class _SessionGroupPropExtractor(object):
```

Comment:
Hmm, I really do think using closures would streamline the code a lot, and it would a bit more like python versus C++ :)  It doesn't have to be in this PR, but maybe a follow-up, or I could do it myself at some point.

Reply:
Fair enough. Let me do it in a follow-up PR?

Reply:
Sure.